### PR TITLE
Push: fix semantics (deprecate --force)

### PR DESCRIPTION
**Pull request by drernie**
# Description
https://app.asana.com/0/1201258662284055/1204391497653719/f

Fixing the put/patch semantics, as I believe this is:

1.  long overdue
2. essential to ensuring '-dir-logical-key' works the way customers will want it to

# TODO

- [ ] Unit tests
- [ ] Automated tests (e.g. Preflight)
- [ ] Confirm that this change meets security best practices and does not violate the security model
- [ ] Documentation
    - [ ] [Python: Run `build.py`](../gendocs/build.py) for new docstrings
    - [ ] Markdown somewhere in docs/**/*.md that explains the feature to end users (said .md files should be linked from SUMMARY.md so they appear on https://docs.quiltdata.com)
    - [ ] Markdown docs for developers
- [ ] [Changelog](CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

### CODE REVIEW

File: .github/workflows/py-ci.yml [hunk 1]
```diff
@@ -163,7 +163,7 @@ jobs:
           - status_reports
           - tabular_preview
           - transcode
-        python-version: ['3.7']
+        python-version: ['3.9']
```
Comment (sir-sigurd) [inline .github/workflows/py-ci.yml:166]: What's the purpose of changes in this file?

**Review by sir-sigurd — CHANGES_REQUESTED** [review]

File: .github/workflows/py-ci.yml (hunk 1 above)
Comment (drernie) [reply .github/workflows/py-ci.yml:166]: This was me messing around trying to figure out why the build was broken. I will start reverting them...

Comment (drernie) [reply .github/workflows/py-ci.yml]: reverted

File: api/python/quilt3/main.py [hunk 2]
```diff
@@ -207,16 +208,30 @@ def _selector_fn_no_copy(*args):
     return False
 
 
-def cmd_push(name, dir, registry, dest, message, meta, workflow, force, dedupe, no_copy):
+def cmd_push(name, dir, registry, dest, message, meta, workflow, force, dedupe, no_copy, request, dir_logical_key):
     if util.PhysicalKey.from_url(util.fix_url(dir)).is_local() and no_copy:
```
Comment (drernie) [inline api/python/quilt3/main.py:212]: I'm not convinced "request" and "dir_logical_key" are the right names.  Anyone have better suggestions?

File: api/python/tests/test_cli.py [hunk 3]
```diff
@@ -54,14 +54,49 @@ def test_push_force(self):
...
+            main.main(('push', '--dir', dir_path, name, '--request', 'patch', '--registry', registry))
+
+            mocked_package_class.browse.assert_called_once_with(name, registry)
+            mocked_package_class.assert_called_once_with()
+            mocked_package = mocked_package_class.return_value
+            mocked_package.set_dir.assert_called_once_with('.', dir_path)
+            mocked_package.push.assert_called_once_with(
+                name, registry=registry, dest=None, message=None, workflow=..., force=False, dedupe=False
+            )
+
```
Comment (drernie) [inline api/python/tests/test_cli.py:99]: I have simplistic integration tests that confirm the CLI is calling the correct API.  I still don't know the right way to ensure we are providing the correct semantics. Any suggestions?

**Comment by drernie** [comment]
Giving up for now. Will fix in quiltcore first, then decide whether/how to back port.

**Pull request closed without merging**